src/core/prometheus.cc

```diff
@@ -43,18 +43,178 @@
+        __builtin_unreachable();
+    }
+};
 
 namespace seastar {
 
 extern seastar::logger seastar_logger;
 
+constexpr std::string_view to_string(seastar::metrics::impl::data_type t) {
+    switch (t) {
+    case seastar::metrics::impl::data_type::GAUGE:
+        return "gauge";
+    case seastar::metrics::impl::data_type::COUNTER:
+    case seastar::metrics::impl::data_type::REAL_COUNTER:
+        return "counter";
+    case seastar::metrics::impl::data_type::HISTOGRAM:
+        return "histogram";
+    case seastar::metrics::impl::data_type::SUMMARY:
+        return "summary";
+    };
+    return "untyped";
+};
+
 namespace prometheus {
 namespace pm = io::prometheus::client;
 
 namespace mi = metrics::impl;
 
 using mi::labels_type;
 
+namespace {
+
+// A wrapper around a std::vector for efficient formatting and appending
+// of text data. Just using a std::vector is much slower.
+// Inspired by log_buf.
+class fmt_buf {
+    char* _current;
+    std::vector<char> buffer;
+private:
+
+    const char* buf_end() const noexcept {
+        return buffer.data() + buffer.size();
+    }
+
+    auto& realloc_buffer_and_append(char c) {
+        auto current_size = size();  // Save size before resizing
+        buffer.resize(buffer.size() * 2);
+        _current = buffer.data() + current_size;
+        *_current++ = c;
+        return *this;
+    }
+
+public:
+    static constexpr size_t initial_size = 1024;
+
+    fmt_buf() {
+        buffer.resize(initial_size);
+        _current = buffer.data();
+    }
+
+    fmt_buf(const fmt_buf&) = delete;
+    fmt_buf(fmt_buf&&) = delete;
+    fmt_buf& operator=(const fmt_buf&) = delete;
+    fmt_buf& operator=(fmt_buf&&) = delete;
+
+    /// Clear the buffer, setting its position back to the start, but does not
+    /// free any buffers (after this called, size is zero, capacity is unchanged).
+    /// Any existing iterators are invalidated.
+    void clear() { _current = data(); }
+
+    /// The amount of data written so far.
+    size_t size() const noexcept { return _current - data(); }
+
+    /// The remaining space
+    size_t remaining() const noexcept { return buf_end() - _current; }
+
+    const char* data() const noexcept { return buffer.data(); }
+    char* data() noexcept { return buffer.data(); }
+
+    auto& append(char c) noexcept {
+        if (_current == buf_end()) [[unlikely]] {
+            return realloc_buffer_and_append(c);
+        }
+        *_current++ = c;
+        return *this;
+    }
+
+    fmt_buf& append(std::string_view str) {
+        if (str.size() <= remaining()) [[likely]] {
+            std::copy(str.begin(), str.end(), _current);
+            _current += str.size();
+        } else {
+            append_slowpath(str);
+        }
+        return *this;
+    }
+
+    fmt_buf& operator<<(std::string_view str) {
+        return append(str);
+    }
+
+    void append_slowpath(std::string_view sv) {
+        for (auto c : sv) {
+            append(c);
+        }
```

Comment:
Why not implement the `realloc_buffer_and_append(std::string_view str)` that would behave similarly to `r.b.and.a(char c)` -- grow the buffers with some reserve and std::copy the string into it?

Reply:
No, not a blocker at all. But a code comment describing this reasoning in few words would be helpful.

Reply:
Comment added in latest push.